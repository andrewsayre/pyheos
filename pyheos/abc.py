"""Define abstract base classes for HEOS."""

from abc import ABC
from typing import Any


class RemoveHeosFieldABC(ABC):
    """Define an abstract base class that removes the 'heos' from dataclass's fields list to prevent serialization."""

    def __post_init__(self, *args: Any, **kwargs: Any) -> None:
        """Post initialize the player."""
        # Prevent the heos instance from being serialized
        fields = self.__dataclass_fields__.copy()  # type: ignore[has-type] # pylint: disable=access-member-before-definition
        del fields["heos"]
        self.__dataclass_fields__ = fields
