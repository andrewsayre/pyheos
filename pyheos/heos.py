"""Define the heos manager module."""
import asyncio
import logging
from typing import Dict, Optional, Sequence

from . import const
from .connection import HeosConnection
from .dispatch import Dispatcher
from .player import HeosPlayer
from .response import HeosResponse
from .source import HeosSource, InputSource

_LOGGER = logging.getLogger(__name__)


class Heos:
    """The Heos class provides access to the CLI API."""

    def __init__(self, host: str, *,
                 timeout: float = const.DEFAULT_TIMEOUT,
                 heart_beat: Optional[float] = const.DEFAULT_HEART_BEAT,
                 all_progress_events=True,
                 dispatcher: Dispatcher = None):
        """Init a new instance of the Heos CLI API."""
        self._connection = HeosConnection(
            self, host, timeout=timeout, heart_beat=heart_beat,
            all_progress_events=all_progress_events)
        self._dispatcher = dispatcher or Dispatcher()
        self._players = {}  # type: Dict[int, HeosPlayer]
        self._players_loaded = False
        self._music_sources = {}  # type: Dict[int, HeosSource]
        self._music_sources_loaded = False

    async def connect(self, *, auto_reconnect=False,
                      reconnect_delay: float = const.DEFAULT_RECONNECT_DELAY):
        """Connect to the CLI."""
        await self._connection.connect(auto_reconnect=auto_reconnect,
                                       reconnect_delay=reconnect_delay)

    async def disconnect(self):
        """Disconnect from the CLI."""
        await self._connection.disconnect()

    async def _handle_event(self, event: HeosResponse) -> bool:
        """Handle a heos event."""
        if event.command == const.EVENT_PLAYERS_CHANGED \
                and self._players_loaded:
            await self.get_players(refresh=True)
        if event.command == const.EVENT_SOURCES_CHANGED \
                and self._music_sources_loaded:
            await self.get_music_sources(refresh=True)
        return True

    async def get_players(self, *, refresh=False) -> Dict[int, HeosPlayer]:
        """Get available players."""
        # get players and pull initial state
        if not self._players or refresh:
            payload = await self._connection.commands.get_players()
            players = {}
            player_data = {}
            for data in payload:
                player = HeosPlayer(self, data)
                players[player.player_id] = player
                player_data[player.player_id] = data
            # Match to existing
            for player_id, player in self._players.items():
                if player_id in players:
                    players.pop(player_id)
                    player.set_available(True)
                    player.from_data(player_data[player_id])
                else:
                    player.set_available(False)

            self._players.update(players)
            # Update all statuses
            await asyncio.gather(*[player.refresh() for player in
                                   self._players.values() if player.available])
            self._players_loaded = True
        return self._players

    async def get_music_sources(self, refresh=True) -> Dict[int, HeosSource]:
        """Get available music sources."""
        if not self._music_sources or refresh:
            payload = await self._connection.commands.get_music_sources()
            self._music_sources.clear()
            for data in payload:
                source = HeosSource(self._connection.commands, data)
                self._music_sources[source.source_id] = source
            self._music_sources_loaded = True
        return self._music_sources

    async def get_input_sources(self) -> Sequence[InputSource]:
        """Get available input sources."""
        payload = await self._connection.commands.browse(
            const.MUSIC_SOURCE_AUX_INPUT)
        sources = [HeosSource(self._connection.commands, item)
                   for item in payload]
        input_sources = []
        for source in sources:
            player_id = source.source_id
            items = await source.browse()
            input_sources.extend(
                [InputSource(player_id, item.name, item.media_id)
                 for item in items])
        return input_sources

    async def get_favorites(self) -> Dict[int, HeosSource]:
        """Get available favorites."""
        payload = await self._connection.commands.browse(
            const.MUSIC_SOURCE_FAVORITES)
        sources = [HeosSource(self._connection.commands, item)
                   for item in payload]
        return {index + 1: source for index, source in enumerate(sources)}

    @property
    def dispatcher(self) -> Dispatcher:
        """Get the dispatcher instance."""
        return self._dispatcher

    @property
    def players(self) -> Dict[int, HeosPlayer]:
        """Get the available players."""
        return self._players

    @property
    def music_sources(self) -> Dict[int, HeosSource]:
        """Get available music sources."""
        return self._music_sources

    @property
    def connection_state(self):
        """Get the state of the connection."""
        return self._connection.state
