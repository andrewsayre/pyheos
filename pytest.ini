[pytest]
log_level=DEBUG
testpaths = tests
norecursedirs = .git
timeout = 60
asyncio_default_fixture_loop_scope = function