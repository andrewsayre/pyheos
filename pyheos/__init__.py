"""pyheos - a library for interacting with HEOS devices."""
from . import const
from .dispatch import Dispatcher
from .heos import Heos
from .player import HeosNowPlayingMedia, HeosPlayer
from .response import CommandError
from .source import HeosSource, InputSource

__all__ = [
    'const',
    'CommandError',
    'Dispatcher',
    'Heos',
    'HeosPlayer',
    'HeosNowPlayingMedia',
    'HeosSource',
    'InputSource'
]
