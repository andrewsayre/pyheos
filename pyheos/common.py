"""Define the common module."""

from typing import cast

from pyheos.const import TARGET_VERSION


def is_supported_version(
    version: str | None, min_version: str = TARGET_VERSION
) -> bool:
    """Check if a version is supported.

    Args:
        version (str): The version to check.
        min_version (str): The minimum version.

    Returns:
        bool: True if the version is supported, False otherwise.
    """
    if version is None:
        return False
    try:
        sem_ver = get_sem_ver(version)
    except ValueError:
        return False
    min_sem_ver = get_sem_ver(min_version)
    for i, ver in enumerate(sem_ver):
        if ver < min_sem_ver[i]:
            return False
        if ver > min_sem_ver[i]:
            return True
    return True


def get_sem_ver(version: str) -> tuple[int, int, int]:
    """Get the semantic version from a string.

    Args:
        version (str): The version string.

    Returns:
        tuple: The semantic version as a tuple.
    """
    return cast(tuple[int, int, int], tuple(map(int, version.split(".", 3))))
