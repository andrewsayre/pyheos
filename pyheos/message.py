"""Define the message module for signals received from HEOS."""

import json
import logging
from dataclasses import dataclass, field
from functools import cached_property
from typing import Any, cast
from urllib.parse import parse_qsl

from pyheos import const

_LOGGER = logging.getLogger(__name__)


@dataclass(frozen=True)
class HeosCommand:
    """Define a HEOS command that is sent to the HEOS device."""

    command: str
    parameters: dict[str, Any] = field(default_factory=dict)

    def __str__(self) -> str:
        """Get a user-readable representation of the command."""
        return self.uri_masked

    def __repr__(self) -> str:
        """Get a debug representation of the command."""
        return self.uri_masked

    @property
    def uri(self) -> str:
        """Get the command as a URI string that can be sent to the controller."""
        return self._get_uri(False)

    @property
    def uri_masked(self) -> str:
        """Get the command as a URI string that has sensitive fields masked."""
        return self._get_uri(True)

    def _get_uri(self, mask: bool = False) -> str:
        """Get the command as a URI string."""
        query_string = (
            f"?{HeosCommand._encode_query(self.parameters, mask=mask)}"
            if self.parameters
            else ""
        )
        return f"{const.BASE_URI}{self.command}{query_string}"

    @classmethod
    def _quote(cls, value: Any) -> str:
        """Quote a string per the CLI specification."""
        return "".join([const.QUOTE_MAP.get(char, char) for char in str(value)])

    @classmethod
    def _encode_query(cls, items: dict[str, Any], *, mask: bool = False) -> str:
        """Encode a dict to query string per CLI specifications."""
        pairs = []
        for key in sorted(items.keys()):
            value = const.MASK if mask and key in const.MASKED_PARAMS else items[key]
            item = f"{key}={HeosCommand._quote(value)}"
            # Ensure 'url' goes last per CLI spec
            if key == const.ATTR_URL:
                pairs.append(item)
            else:
                pairs.insert(0, item)
        return "&".join(pairs)


@dataclass(frozen=True)
class HeosMessage:
    """Lower a message received from a HEOS device. This is a lower level class used internally."""

    raw_message: str

    def __str__(self) -> str:
        """Get a user-readable representation of the message."""
        return str(self.raw_message)

    def __repr__(self) -> str:
        """Get a debug representation of the message."""
        return str(self.raw_message)

    @cached_property
    def container(self) -> dict[str, Any]:
        """Get the entire message as a dictionary."""
        return cast(dict[str, Any], json.loads(self.raw_message))

    @cached_property
    def heos(self) -> dict[str, Any]:
        """Get the HEOS section as a dictionary."""
        return cast(dict[str, Any], self.container["heos"])

    @cached_property
    def payload(self) -> dict[str, Any] | list[Any] | None:
        """Get the payload section as a dictionary."""
        return self.container.get("payload")

    @cached_property
    def command(self) -> str:
        """Get the command the message is referring to."""
        return str(self.heos["command"])

    @cached_property
    def message(self) -> dict[str, str]:
        """Get the message parameters as a dictionary. If not present in the message, an empty dictionary is returned."""
        if raw_message := self.heos.get("message"):
            return dict(parse_qsl(raw_message, keep_blank_values=True))
        return {}

    @cached_property
    def is_under_process(self) -> bool:
        """Return True if the message represents a command under process, otherwise False."""
        return "command under process" in self.message

    @cached_property
    def is_event(self) -> bool:
        """Return True if the message is an event, otherwise False."""
        return self.command.startswith("event/")

    @cached_property
    def result(self) -> bool:
        """Return True if the message represents a successful command. If not present in the response, True is returned."""
        if raw_result := self.heos.get("result"):
            return str(raw_result) == "success"
        return True

    def get_message_value(self, key: str) -> str:
        """Get a message parameter as a string."""
        assert self.message is not None
        if key not in self.message:
            raise KeyError(f"Key '{key}' not found in message parameters.")
        return self.message[key]

    def get_message_value_float(self, key: str) -> float:
        """Get a message parameter as a float."""
        return float(self.get_message_value(key))

    def get_message_value_int(self, key: str) -> int:
        """Get a message parameter as an integer.

        Convert to float first because it may contain a decimal point.
        """
        return int(self.get_message_value_float(key))
