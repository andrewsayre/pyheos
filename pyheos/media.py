""" "Define the media module."""

from collections.abc import Sequence
from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any, Optional, cast

from pyheos import const
from pyheos.message import HeosMessage

if TYPE_CHECKING:
    from . import Heos


@dataclass(init=False)
class Media:
    """
    Define a base media item.

    Do not instantiate directly. Use either MediaMusicSource or MediaItem.
    """

    source_id: int
    name: str
    type: const.MediaType
    image_url: str = field(repr=False)
    heos: Optional["Heos"] = field(repr=False, hash=False, compare=False)


@dataclass
class MediaMusicSource(Media):
    """
    Define a music source.

    A Music Source is a top-level music source, only returned from get_music_sources.
    """

    available: bool
    service_username: str | None

    @classmethod
    def from_data(
        cls,
        data: dict[str, Any],
        heos: Optional["Heos"] = None,
    ) -> "MediaMusicSource":
        """Create a new instance from the provided data."""
        return cls(
            source_id=int(data[const.ATTR_SOURCE_ID]),
            name=data[const.ATTR_NAME],
            type=const.MediaType(data[const.ATTR_TYPE]),
            image_url=data[const.ATTR_IMAGE_URL],
            available=data[const.ATTR_AVAILABLE] == const.VALUE_TRUE,
            service_username=data.get(const.ATTR_SERVICE_USER_NAME),
            heos=heos,
        )

    def clone(self) -> "MediaMusicSource":
        """Create a new instance from the current instance."""
        return MediaMusicSource(
            source_id=self.source_id,
            name=self.name,
            type=self.type,
            image_url=self.image_url,
            available=self.available,
            service_username=self.service_username,
            heos=self.heos,
        )

    async def browse(self) -> "BrowseResult":
        """Browse the contents of this source.

        Returns:
            A BrowseResult instance containing the items in this source."""
        assert self.heos, "Heos instance not set"
        return await self.heos.browse(self.source_id)


@dataclass
class MediaItem(Media):
    """Define a playable media item."""

    playable: bool
    browsable: bool
    container_id: str | None = None
    media_id: str | None = None
    artist: str | None = None
    album: str | None = None
    album_id: str | None = None

    @classmethod
    def from_data(
        cls,
        data: dict[str, Any],
        source_id: int | None = None,
        container_id: str | None = None,
        heos: Optional["Heos"] = None,
    ) -> "MediaItem":
        """Create a new instance from the provided data."""

        # Ensure we have a source_id
        if const.ATTR_SOURCE_ID not in data and not source_id:
            raise ValueError("'source_id' is required when not present in 'data'")
        new_source_id = int(data.get(const.ATTR_SOURCE_ID, source_id))
        # Items is browsable if is a media source, or if it is a container
        new_browseable = (
            const.ATTR_SOURCE_ID in data
            or data.get(const.ATTR_CONTAINER) == const.VALUE_YES
        )

        return cls(
            source_id=new_source_id,
            container_id=data.get(const.ATTR_CONTAINER_ID, container_id),
            browsable=new_browseable,
            name=data[const.ATTR_NAME],
            type=const.MediaType(data[const.ATTR_TYPE]),
            image_url=data[const.ATTR_IMAGE_URL],
            playable=data.get(const.ATTR_PLAYABLE) == const.VALUE_YES,
            media_id=data.get(const.ATTR_MEDIA_ID),
            artist=data.get(const.ATTR_ARTIST),
            album=data.get(const.ATTR_ALBUM),
            album_id=data.get(const.ATTR_ALBUM_ID),
            heos=heos,
        )

    def clone(self) -> "MediaItem":
        return MediaItem(
            source_id=self.source_id,
            name=self.name,
            type=self.type,
            image_url=self.image_url,
            playable=self.playable,
            browsable=self.browsable,
            container_id=self.container_id,
            media_id=self.media_id,
            artist=self.artist,
            album=self.album,
            album_id=self.album_id,
            heos=self.heos,
        )

    async def browse(
        self,
        range_start: int | None = None,
        range_end: int | None = None,
    ) -> "BrowseResult":
        """Browse the contents of the current media item (source or container).

        Args:
            range_start: The index of the first item to return. Both range_start and range_end must be provided to return a range of items.
            range_end: The index of the last item to return. Both range_start and range_end must be provided to return a range of items.
        Returns:
            A BrowseResult instance containing the items in this media item (source or container)."""
        assert self.heos, "Heos instance not set"
        return await self.heos.browse_media(self, range_start, range_end)

    async def play_media(
        self,
        player_id: int,
        add_criteria: const.AddCriteriaType = const.AddCriteriaType.PLAY_NOW,
    ) -> None:
        """Play this media item on the specified player.

        Args:
            player_id: The id of the player to play on.
        """
        assert self.heos, "Heos instance not set"
        await self.heos.play_media(player_id, self, add_criteria)


@dataclass
class BrowseResult:
    """Define the result of a browse operation."""

    count: int
    returned: int
    source_id: int
    items: Sequence[MediaItem] = field(repr=False, hash=False, compare=False)
    container_id: str | None = None
    heos: Optional["Heos"] = field(repr=False, hash=False, compare=False, default=None)

    @classmethod
    def from_data(
        cls, message: HeosMessage, heos: Optional["Heos"] = None
    ) -> "BrowseResult":
        """Create a new instance from the provided data."""
        source_id = message.get_message_value_int(const.ATTR_SOURCE_ID)
        container_id = message.message.get(const.ATTR_CONTAINER_ID)

        return cls(
            count=message.get_message_value_int(const.ATTR_COUNT),
            returned=message.get_message_value_int(const.ATTR_RETURNED),
            source_id=source_id,
            container_id=container_id,
            items=list(
                [
                    MediaItem.from_data(item, source_id, container_id, heos)
                    for item in cast(Sequence[dict], message.payload)
                ]
            ),
            heos=heos,
        )
