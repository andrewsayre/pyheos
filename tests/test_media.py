"""Define tests for browsing media."""

import re
from unittest.mock import Mock

import pytest

from pyheos import command, const
from pyheos.heos import Heos
from pyheos.media import BrowseResult, MediaItem, MediaMusicSource
from pyheos.message import HeosMessage
from tests import calls_command
from tests.common import MediaItems


async def test_media_music_source_from_data() -> None:
    """Test creating a media music source from data."""
    data = {
        const.ATTR_NAME: "Pandora",
        const.ATTR_IMAGE_URL: "https://production.ws.skyegloup.com:443/media/images/service/logos/pandora.png",
        const.ATTR_TYPE: const.MediaType.MUSIC_SERVICE,
        const.ATTR_SOURCE_ID: 1,
        const.ATTR_AVAILABLE: const.VALUE_TRUE,
        const.ATTR_SERVICE_USER_NAME: "test@test.com",
    }

    source = MediaMusicSource.from_data(data)

    assert source.name == data[const.ATTR_NAME]
    assert source.image_url == data[const.ATTR_IMAGE_URL]
    assert source.type == const.MediaType.MUSIC_SERVICE
    assert source.source_id == data[const.ATTR_SOURCE_ID]
    assert source.available
    assert source.service_username == data[const.ATTR_SERVICE_USER_NAME]
    with pytest.raises(
        AssertionError,
        match="Heos instance not set",
    ):
        await source.browse()


@calls_command(
    "browse.browse_favorites", {const.ATTR_SOURCE_ID: const.MUSIC_SOURCE_FAVORITES}
)
async def test_media_music_source_browse(
    media_music_source: MediaMusicSource,
) -> None:
    """Test browsing a media music source."""
    result = await media_music_source.browse()

    assert result.returned == 3
    assert result.source_id == const.MUSIC_SOURCE_FAVORITES
    # further testing of the result is done in test_browse_result_from_data


async def test_browse_result_from_data() -> None:
    """Test creating a browse result from data."""
    heos = Mock(Heos)
    message = HeosMessage(
        command.COMMAND_BROWSE_BROWSE,
        True,
        {const.ATTR_SOURCE_ID: "1025", const.ATTR_RETURNED: "1", const.ATTR_COUNT: "1"},
        [
            {
                const.ATTR_CONTAINER: const.VALUE_YES,
                const.ATTR_TYPE: str(const.MediaType.PLAYLIST),
                const.ATTR_CONTAINER_ID: "171566",
                const.ATTR_PLAYABLE: const.VALUE_YES,
                const.ATTR_NAME: "Rockin Songs",
                const.ATTR_IMAGE_URL: "",
            }
        ],
    )

    result = BrowseResult.from_data(message, heos)

    assert result.returned == 1
    assert result.count == 1
    assert result.source_id == 1025
    assert result.heos == heos
    assert len(result.items) == 1
    item = result.items[0]
    assert item.heos == heos


async def test_media_item_from_data() -> None:
    """Test creating a MediaItem from data."""
    source_id = 1
    container_id = "My Music"
    data = {
        const.ATTR_NAME: "Imaginary Parties",
        const.ATTR_IMAGE_URL: "http://resources.wimpmusic.com/images/7e7bacc1/3e75/4761/a822/9342239edfa0/640x640.jpg",
        const.ATTR_TYPE: str(const.MediaType.SONG),
        const.ATTR_CONTAINER: const.VALUE_NO,
        const.ATTR_MEDIA_ID: "78374741",
        const.ATTR_ARTIST: "Superfruit",
        const.ATTR_ALBUM: "Future Friends",
        const.ATTR_ALBUM_ID: "78374740",
        const.ATTR_PLAYABLE: const.VALUE_YES,
    }

    source = MediaItem.from_data(data, source_id, container_id)

    assert source.name == data[const.ATTR_NAME]
    assert source.image_url == data[const.ATTR_IMAGE_URL]
    assert source.type == const.MediaType.SONG
    assert source.container_id == container_id
    assert source.source_id == source_id
    assert source.playable is True
    assert source.browsable is False
    assert source.album == data[const.ATTR_ALBUM]
    assert source.artist == data[const.ATTR_ARTIST]
    assert source.album_id == data[const.ATTR_ALBUM_ID]
    assert source.media_id == data[const.ATTR_MEDIA_ID]
    with pytest.raises(
        AssertionError,
        match="Heos instance not set",
    ):
        await source.browse()
    with pytest.raises(
        AssertionError,
        match="Heos instance not set",
    ):
        await source.play_media(1)


async def test_media_item_from_data_source_id_not_present_raises() -> None:
    """Test creating a MediaItem from data."""
    data = {
        const.ATTR_NAME: "Video",
        const.ATTR_IMAGE_URL: "",
        const.ATTR_TYPE: str(const.MediaType.CONTAINER),
        const.ATTR_CONTAINER: const.VALUE_YES,
        const.ATTR_CONTAINER_ID: "94467912-bd40-4d2f-ad25-7b8423f7b05a",
    }

    with pytest.raises(
        ValueError,
        match=re.escape("'source_id' is required when not present in 'data'"),
    ):
        MediaItem.from_data(data)


async def test_media_item_from_data_source() -> None:
    """Test creating a MediaItem from data."""
    data = {
        const.ATTR_NAME: "Plex Media Server",
        const.ATTR_IMAGE_URL: "https://production.ws.skyegloup.com:443/media/images/service/logos/musicsource_logo_servers.png",
        const.ATTR_TYPE: str(const.MediaType.HEOS_SERVER),
        const.ATTR_SOURCE_ID: 123456789,
    }

    source = MediaItem.from_data(data)

    assert source.name == data[const.ATTR_NAME]
    assert source.image_url == data[const.ATTR_IMAGE_URL]
    assert source.type == const.MediaType.HEOS_SERVER
    assert source.source_id == data[const.ATTR_SOURCE_ID]
    assert source.container_id is None
    assert source.playable is False
    assert source.browsable is True
    assert source.album is None
    assert source.artist is None
    assert source.album_id is None
    assert source.media_id is None


async def test_media_item_from_data_container() -> None:
    """Test creating a MediaItem from data."""
    source_id = 123456789
    data = {
        const.ATTR_NAME: "Video",
        const.ATTR_IMAGE_URL: "",
        const.ATTR_TYPE: str(const.MediaType.CONTAINER),
        const.ATTR_CONTAINER: const.VALUE_YES,
        const.ATTR_CONTAINER_ID: "94467912-bd40-4d2f-ad25-7b8423f7b05a",
    }

    source = MediaItem.from_data(data, source_id)

    assert source.name == data[const.ATTR_NAME]
    assert source.image_url == data[const.ATTR_IMAGE_URL]
    assert source.type == const.MediaType.CONTAINER
    assert source.container_id == data[const.ATTR_CONTAINER_ID]
    assert source.source_id == source_id
    assert source.playable is False
    assert source.browsable is True
    assert source.album is None
    assert source.artist is None
    assert source.album_id is None
    assert source.media_id is None


@calls_command(
    "browse.browse_heos_drive", {const.ATTR_SOURCE_ID: MediaItems.DEVICE.source_id}
)
async def test_media_item_browse(media_item_device: MediaItem) -> None:
    """Test browsing a media music source."""
    result = await media_item_device.browse()

    assert result.container_id is None
    assert result.source_id == media_item_device.source_id
    assert result.returned == 8
    assert result.count == 8
    assert len(result.items) == 8


@calls_command(
    "browse.add_to_queue_track",
    {
        const.ATTR_PLAYER_ID: 1,
        const.ATTR_SOURCE_ID: MediaItems.SONG.source_id,
        const.ATTR_CONTAINER_ID: MediaItems.SONG.container_id,
        const.ATTR_MEDIA_ID: MediaItems.SONG.media_id,
        const.ATTR_ADD_CRITERIA_ID: const.AddCriteriaType.REPLACE_AND_PLAY,
    },
    add_command_under_process=True,
)
async def test_media_item_play(media_item_song: MediaItem) -> None:
    """Test playing a media music source."""
    await media_item_song.play_media(1, const.AddCriteriaType.REPLACE_AND_PLAY)
