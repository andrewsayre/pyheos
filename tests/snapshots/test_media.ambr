# serializer version: 1
# name: test_browse_result_from_data
  dict({
    'container_id': None,
    'count': 1,
    'items': list([
      dict({
        'album': None,
        'album_id': None,
        'artist': None,
        'browsable': True,
        'container_id': '171566',
        'image_url': '',
        'media_id': None,
        'name': 'Rockin Songs',
        'playable': True,
        'source_id': 1025,
        'type': <MediaType.PLAYLIST: 'playlist'>,
      }),
    ]),
    'options': list([
    ]),
    'returned': 1,
    'source_id': 1025,
  })
# ---
# name: test_media_item_browse
  dict({
    'container_id': None,
    'count': 8,
    'items': list([
      dict({
        'album': None,
        'album_id': None,
        'artist': None,
        'browsable': False,
        'container_id': None,
        'image_url': '',
        'media_id': 'inputs/line_in_1',
        'name': 'HEOS Drive - Line In 1',
        'playable': True,
        'source_id': -263109739,
        'type': <MediaType.STATION: 'station'>,
      }),
      dict({
        'album': None,
        'album_id': None,
        'artist': None,
        'browsable': False,
        'container_id': None,
        'image_url': '',
        'media_id': 'inputs/line_in_2',
        'name': 'HEOS Drive - Line In 2',
        'playable': True,
        'source_id': -263109739,
        'type': <MediaType.STATION: 'station'>,
      }),
      dict({
        'album': None,
        'album_id': None,
        'artist': None,
        'browsable': False,
        'container_id': None,
        'image_url': '',
        'media_id': 'inputs/line_in_3',
        'name': 'HEOS Drive - Line In 3',
        'playable': True,
        'source_id': -263109739,
        'type': <MediaType.STATION: 'station'>,
      }),
      dict({
        'album': None,
        'album_id': None,
        'artist': None,
        'browsable': False,
        'container_id': None,
        'image_url': '',
        'media_id': 'inputs/line_in_4',
        'name': 'HEOS Drive - Line In 4',
        'playable': True,
        'source_id': -263109739,
        'type': <MediaType.STATION: 'station'>,
      }),
      dict({
        'album': None,
        'album_id': None,
        'artist': None,
        'browsable': False,
        'container_id': None,
        'image_url': '',
        'media_id': 'inputs/optical_in_1',
        'name': 'HEOS Drive - Optical In 1',
        'playable': True,
        'source_id': -263109739,
        'type': <MediaType.STATION: 'station'>,
      }),
      dict({
        'album': None,
        'album_id': None,
        'artist': None,
        'browsable': False,
        'container_id': None,
        'image_url': '',
        'media_id': 'inputs/optical_in_2',
        'name': 'HEOS Drive - Optical In 2',
        'playable': True,
        'source_id': -263109739,
        'type': <MediaType.STATION: 'station'>,
      }),
      dict({
        'album': None,
        'album_id': None,
        'artist': None,
        'browsable': False,
        'container_id': None,
        'image_url': '',
        'media_id': 'inputs/coax_in_1',
        'name': 'HEOS Drive - Coaxial In 1',
        'playable': True,
        'source_id': -263109739,
        'type': <MediaType.STATION: 'station'>,
      }),
      dict({
        'album': None,
        'album_id': None,
        'artist': None,
        'browsable': False,
        'container_id': None,
        'image_url': '',
        'media_id': 'inputs/coax_in_2',
        'name': 'HEOS Drive - Coaxial In 2',
        'playable': True,
        'source_id': -263109739,
        'type': <MediaType.STATION: 'station'>,
      }),
    ]),
    'options': list([
    ]),
    'returned': 8,
    'source_id': -263109739,
  })
# ---
# name: test_media_item_from_data
  dict({
    'album': 'Future Friends',
    'album_id': '78374740',
    'artist': 'Superfruit',
    'browsable': False,
    'container_id': 'My Music',
    'image_url': 'http://resources.wimpmusic.com/images/7e7bacc1/3e75/4761/a822/9342239edfa0/640x640.jpg',
    'media_id': '78374741',
    'name': 'Imaginary Parties',
    'playable': True,
    'source_id': 1,
    'type': <MediaType.SONG: 'song'>,
  })
# ---
# name: test_media_item_from_data_container
  dict({
    'album': None,
    'album_id': None,
    'artist': None,
    'browsable': True,
    'container_id': '94467912-bd40-4d2f-ad25-7b8423f7b05a',
    'image_url': '',
    'media_id': None,
    'name': 'Video',
    'playable': False,
    'source_id': 123456789,
    'type': <MediaType.CONTAINER: 'container'>,
  })
# ---
# name: test_media_item_from_data_source
  dict({
    'album': None,
    'album_id': None,
    'artist': None,
    'browsable': True,
    'container_id': None,
    'image_url': 'https://production.ws.skyegloup.com:443/media/images/service/logos/musicsource_logo_servers.png',
    'media_id': None,
    'name': 'Plex Media Server',
    'playable': False,
    'source_id': 123456789,
    'type': <MediaType.HEOS_SERVER: 'heos_server'>,
  })
# ---
# name: test_media_music_source_browse
  dict({
    'container_id': None,
    'count': 3,
    'items': list([
      dict({
        'album': None,
        'album_id': None,
        'artist': None,
        'browsable': False,
        'container_id': None,
        'image_url': 'http://mediaserver-cont-ch1-1-v4v6.pandora.com/images/public/devicead/t/r/a/m/daartpralbumart_500W_500H.jpg',
        'media_id': '3790855220637622543',
        'name': 'Thumbprint Radio',
        'playable': True,
        'source_id': 1028,
        'type': <MediaType.STATION: 'station'>,
      }),
      dict({
        'album': None,
        'album_id': None,
        'artist': None,
        'browsable': False,
        'container_id': None,
        'image_url': 'http://cdn-radiotime-logos.tunein.com/s7332q.png',
        'media_id': 's7332',
        'name': '99.5 | Classical MPR (Classical Music)',
        'playable': True,
        'source_id': 1028,
        'type': <MediaType.STATION: 'station'>,
      }),
      dict({
        'album': None,
        'album_id': None,
        'artist': None,
        'browsable': False,
        'container_id': None,
        'image_url': 'http://mediaserver-cont-dc6-2-v4v6.pandora.com/images/public/int/4/0/8/2/00602577432804_500W_500H.jpg',
        'media_id': '1935916853323522319',
        'name': "Today's Hits Radio",
        'playable': True,
        'source_id': 1028,
        'type': <MediaType.STATION: 'station'>,
      }),
    ]),
    'options': list([
      dict({
        'context': 'browse',
        'id': 20,
        'name': 'Remove from HEOS Favorites',
      }),
    ]),
    'returned': 3,
    'source_id': 1028,
  })
# ---
# name: test_media_music_source_from_data
  dict({
    'available': True,
    'image_url': 'https://production.ws.skyegloup.com:443/media/images/service/logos/pandora.png',
    'name': 'Pandora',
    'service_username': 'test@test.com',
    'source_id': 1,
    'type': <MediaType.MUSIC_SERVICE: 'music_service'>,
  })
# ---
# name: test_refresh
  dict({
    'available': True,
    'image_url': 'https://production.ws.skyegloup.com:443/media/images/service/logos/pandora.png',
    'name': 'Pandora',
    'service_username': 'email@email.com',
    'source_id': 1,
    'type': <MediaType.MUSIC_SERVICE: 'music_service'>,
  })
# ---
